plutus-core/plutus-ir/test/TypecheckSpec.hs

```diff
@@ -0,0 +1,141 @@
+  , testProperty "beta pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_pure beta)
+  , testProperty "commuteFnWithConst pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_pure commuteFnWithConst)
+  -- non-pure passes
+  -- FIXME
+  , expectFail $ testProperty "the whole simplifier pass" $
+      withMaxSuccess (factor*3000) prop_typecheck
+  , testProperty "strictifyBindings (Builtin Variant 1) pass" $
+      withMaxSuccess (factor*3000) (prop_typecheck_strictifyBindings DefaultFunSemanticsVariant1)
+  , testProperty "strictifyBindings (Builtin Variant 2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_strictifyBindings DefaultFunSemanticsVariant2)
+  , testProperty "evaluateBuiltins (Builtin Variant1) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_evaluateBuiltins DefaultFunSemanticsVariant1)
+  , testProperty "evaluateBuiltins (Builtin Variant2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_evaluateBuiltins DefaultFunSemanticsVariant2)
+  -- FIXME
+  , expectFail $ testProperty "inline (Builtin Variant1) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_inline DefaultFunSemanticsVariant1)
+  -- FIXME
+  , expectFail $ testProperty "inline (Builtin Variant2) pass" $
+  withMaxSuccess (factor*3000) (prop_typecheck_inline DefaultFunSemanticsVariant2)
+  ]
+
+-- Convert Either Error () to Either String () to match with the Testable (Either String ())
+-- instance.
+convertToEitherString :: Either (Error DefaultUni DefaultFun ()) ()
+  -> Either String ()
+convertToEitherString = \case
+  Left err -> Left $ show err
+  Right () -> Right ()
+
+-- | Typechecking a term after one of the pure passes.
+tyCheckPure ::
+  -- | The pure simplification pass.
+  (Term TyName Name DefaultUni DefaultFun () -> Term TyName Name DefaultUni DefaultFun ())
+  -> Term TyName Name DefaultUni DefaultFun ()
+  -> Either (Error DefaultUni DefaultFun ()) ()
+tyCheckPure pass tm = do
+  config <- getDefTypeCheckConfig ()
+  inferredType <- runQuoteT $ inferType config tm
+  runQuoteT $ checkType config () (pass tm) inferredType
+
+prop_typecheck_pure ::
+  (Term TyName Name DefaultUni DefaultFun ()
+  -> Term TyName Name DefaultUni DefaultFun ())
+  -> Property
+prop_typecheck_pure pass =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) -> do
+  convertToEitherString $ tyCheckPure pass tm
+
+-- | Typechecking a term after a `PlutusIR.Compiler.simplifyTerm` pass.
+prop_typecheck :: Property
+prop_typecheck =
+  -- generate type and term in debug mode for easier debugging
+  forAllDoc "ty,tm" genTypeAndTermDebug_ (const []) $ \ (_ty, tm) ->
+    convertToEitherString $ do
+    config <- getDefTypeCheckConfig ()
+    plcConfig <- PLC.getDefTypeCheckConfig ()
+    inferredType <- runQuoteT $ inferType config tm
+    case runReaderT (runQuoteT (simplifyTerm (original tm))) (toDefaultCompilationCtx plcConfig) of
```

Comment:
I'd have thought we'd want to start with the individual passes, rather than running the whole simplify loop?

Reply:
Sorry I'm not sure what you mean. Did you want me to remove the test for the whole loop? Why?

Reply:
We don't test the "simplifier loop as a whole" anywhere else. Also it seems pretty clear to me that if all the individual passes in the simplifier preserve typechecking, then so must the simplifier. I guess it's not bad per se, but if I was going to do an "integration test" for the compiler like this, I'd probably run the _whole_ compiler with pedantic mode on (which typechecks at every step).

Reply:
I don't see a strong reason to remove it. I think it makes sense to test more increments when possible. Although I don't mind removing it if it annoys you.